From 117f64885904f743d1f04c5ddbbd39aab5a56e6d Mon Sep 17 00:00:00 2001
From: Josh Blum <josh@joshknows.com>
Date: Fri, 14 Apr 2017 11:02:00 -0500
Subject: [PATCH 89/96] soapy: support set/getFrequencyCorrection() API

backwards compatible changes with #ifdef
set/get_freq_corr() call directly into the SoapySDR
equivalent when supported by the API version.
---
 lib/soapy/soapy_sink_c.cc   | 8 ++++++++
 lib/soapy/soapy_source_c.cc | 8 ++++++++
 2 files changed, 16 insertions(+)

diff --git a/lib/soapy/soapy_sink_c.cc b/lib/soapy/soapy_sink_c.cc
index 29defea..07cf069 100644
--- a/lib/soapy/soapy_sink_c.cc
+++ b/lib/soapy/soapy_sink_c.cc
@@ -171,22 +171,30 @@ double soapy_sink_c::get_center_freq( size_t chan)
 
 double soapy_sink_c::set_freq_corr( double ppm, size_t chan)
 {
+    #ifdef SOAPY_SDR_API_HAS_FREQUENCY_CORRECTION_API
+    _device->setFrequencyCorrection(SOAPY_SDR_TX, chan, ppm);
+    #else
     std::vector<std::string> components = _device->listFrequencies(SOAPY_SDR_TX, chan);
     if (std::find(components.begin(), components.end(), "COOR") != components.end())
     {
         _device->setFrequency(SOAPY_SDR_TX, chan, "CORR", ppm);
     }
+    #endif
     return this->get_freq_corr(chan);
 }
 
 double soapy_sink_c::get_freq_corr( size_t chan)
 {
+    #ifdef SOAPY_SDR_API_HAS_FREQUENCY_CORRECTION_API
+    return _device->getFrequencyCorrection(SOAPY_SDR_TX, chan);
+    #else
     std::vector<std::string> components = _device->listFrequencies(SOAPY_SDR_TX, chan);
     if (std::find(components.begin(), components.end(), "COOR") != components.end())
     {
         return _device->getFrequency(SOAPY_SDR_TX, chan, "CORR");
     }
     return 0.0;
+    #endif
 }
 
 std::vector<std::string> soapy_sink_c::get_gain_names( size_t chan)
diff --git a/lib/soapy/soapy_source_c.cc b/lib/soapy/soapy_source_c.cc
index a1a1885..c780966 100644
--- a/lib/soapy/soapy_source_c.cc
+++ b/lib/soapy/soapy_source_c.cc
@@ -172,22 +172,30 @@ double soapy_source_c::get_center_freq( size_t chan )
 
 double soapy_source_c::set_freq_corr( double ppm, size_t chan )
 {
+    #ifdef SOAPY_SDR_API_HAS_FREQUENCY_CORRECTION_API
+    _device->setFrequencyCorrection(SOAPY_SDR_RX, chan, ppm);
+    #else
     std::vector<std::string> components = _device->listFrequencies(SOAPY_SDR_RX, chan);
     if (std::find(components.begin(), components.end(), "COOR") != components.end())
     {
         _device->setFrequency(SOAPY_SDR_RX, chan, "CORR", ppm);
     }
+    #endif
     return this->get_freq_corr(chan);
 }
 
 double soapy_source_c::get_freq_corr( size_t chan )
 {
+    #ifdef SOAPY_SDR_API_HAS_FREQUENCY_CORRECTION_API
+    return _device->getFrequencyCorrection(SOAPY_SDR_RX, chan);
+    #else
     std::vector<std::string> components = _device->listFrequencies(SOAPY_SDR_RX, chan);
     if (std::find(components.begin(), components.end(), "COOR") != components.end())
     {
         return _device->getFrequency(SOAPY_SDR_RX, chan, "CORR");
     }
     return 0.0;
+    #endif
 }
 
 std::vector<std::string> soapy_source_c::get_gain_names( size_t chan )
-- 
2.11.0

